### DISCUSSION TITLE
Toggle between Miryoku and my custom keymap?

### DISCUSSION THREAD

**Discussion by azzamsa** [manna-harbour/miryoku#346] [Q&A] [answered]
Hi, I am using the Miryoku keymap for all my keyboards.

I now have a Sofle keyboard in hand, and I want to use it for gaming and typing.
How do I add a button to toggle between miryoku and my gaming keymap?

What is the easiest approach? Should I add something like MIRYOKU_ALTERNATIVES_TAP_KEYPAD, [here](https://github.com/manna-harbour/miryoku_qmk/blob/df1690ad257e7b0d1bf9450258f7baebfee7be9a/users/manna-harbour_miryoku/miryoku_babel/miryoku_layer_alternatives.h#L240) ?

Then build using something like `qmk compile -c -kb sofle -km manna-harbour_miryoku -e MIRYOKU_TAP=KEYPAD`?

My gaming keymap looks like this:

```c
#include QMK_KEYBOARD_H

const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
    [0] = LAYOUT(
        KC_5,  KC_ESC,  KC_1,     KC_2,    KC_3,    KC_4,                          XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
        KC_T,  KC_TAB,  KC_Q,     KC_W,    KC_E,    KC_R,                          XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
        KC_G,  KC_LSFT, KC_A,     KC_S,    KC_D,    KC_F,                          XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
        KC_B,  KC_LCTL, KC_Z,     KC_X,    KC_C,    KC_V,    XXXXXXX,    XXXXXXX,  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
                        KC_PSCR,  KC_ENT, OSL(2),  KC_SPC,  OSL(1),     XXXXXXX,  XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX
    ),

    [1] = LAYOUT(
        KC_0,     KC_GRV,   KC_6,    KC_7,     KC_8,     KC_9,                         XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
        KC_P,     _______,  KC_Y,    KC_U,     KC_I,     KC_O,                         XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
        KC_SCLN,  _______,  KC_H,    KC_J,     KC_K,     KC_L,                         XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
        KC_SLSH,  _______,  KC_N,    KC_M,     KC_COMM,  KC_DOT,  XXXXXXX,    XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
                            KC_F12,  _______,  XXXXXXX,  KC_DEL,  _______,    XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX
    ),

...
```

**Comment by manna-harbour** [comment, marked as answer]
You can add and substitute layers using https://github.com/manna-harbour/miryoku/discussions/85. You don't need to edit anything under `miryoku_babel/`.

The Tap and Extra layers are available as alternative base layers and can be substituted. Additional layers can be added and can be reached from the the substituted base layer. You could also add extra base layers, but you'd need to add bindings to Base to reach them.

To map the extra keys, also see https://github.com/manna-harbour/miryoku/discussions/85. Custom mappings can be substituted for all layers or per-layer.

**Reply by manna-harbour** [reply]
Alternatively, for QMK, you could enable https://docs.qmk.fm/features/split_keyboard#handedness-by-eeprom, maintain a custom gaming keymap separately, flash it to the other side, and switch the USB cable to switch keymaps. Both builds should be built from the same QMK source and keyboard configuration.
